### PR TITLE
Fix softmax behavior to not cast up the accumulator if no output dtype is specified

### PR DESCRIPTION
**Pull request by nswamy**
## Description ##
Softmax operator was recently changed to cast the accumulator type to higher precision data type(in #14098) than the input type regardless of the whether the `dtype` of the output is passed to the operator or not. This not only breaks backward compatibility but also the  increases loss when training a rnn language model with float32 input data type. Details are https://github.com/apache/incubator-mxnet/issues/14722. 

This PR changes the accumulator type only if the `dtype` has been passed and uses the `dtype` passed for accumulation.  

@eric-haibin-lin 

### Essentials ###
Please feel free to remove inapplicable items for your PR.
- [ ] The PR title starts with [MXNET-$JIRA_ID], where $JIRA_ID refers to the relevant [JIRA issue](https://issues.apache.org/jira/projects/MXNET/issues) created (except PRs with tiny changes)
- [ ] Changes are complete (i.e. I finished coding on this PR)
- [ ] All changes have test coverage:
- Unit tests are added for small changes to verify correctness (e.g. adding a new operator)
- Nightly tests are added for complicated/long-running ones (e.g. changing distributed kvstore)
- Build tests will be added for build configuration changes (e.g. adding a new build option with NCCL)
- [ ] Code is well-documented: 
- For user-facing API changes, API doc string has been updated. 
- For new C++ functions in header files, their functionalities and arguments are documented. 
- For new examples, README.md is added to explain the what the example does, the source of the dataset, expected performance on test set and reference to the original paper if applicable
- Check the API doc at http://mxnet-ci-doc.s3-accelerate.dualstack.amazonaws.com/PR-$PR_ID/$BUILD_ID/index.html
- [ ] To the my best knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### Changes ###
- [ ] Feature1, tests, (and when applicable, API doc)
- [ ] Feature2, tests, (and when applicable, API doc)

## Comments ##
- If this change is a backward incompatible change, why must this change be made.
- Interesting edge cases to note here

### CODE REVIEW

File: src/operator/nn/softmax-inl.h [hunk 1]
```diff
@@ -310,9 +311,9 @@ struct SoftmaxParam : public dmlc::Parameter<SoftmaxParam> {
   }
 };
 
-static inline bool softmax_has_dtype_override(const nnvm::NodeAttrs& attrs) {
+static inline int sofmtax_dtype_param(const nnvm::NodeAttrs &attrs) {
```
Comment (eric-haibin-lin) [inline src/operator/nn/softmax-inl.h]: typo? softmax

File: src/operator/nn/softmax-inl.h [hunk 2]
```diff
@@ -442,23 +449,28 @@ void SoftmaxGradCompute(const nnvm::NodeAttrs& attrs,
...
-              inputs[0].dptr<OType>(), outputs[0].dptr<DType>(),
-              shape.get<2>(), axis, static_cast<DType>(temperature));
-        } else {
-          SoftmaxGrad<OP1, OP2, Req, negate, AType>(
-              ctx.get_stream<xpu>(), inputs[out_idx].dptr<OType>(),
-              inputs[0].dptr<OType>(), outputs[0].dptr<DType>(),
-              shape.get<3>(), axis, static_cast<DType>(temperature));
-        }
-      });
+  int out_idx = sofmtax_dtype_param(attrs) != -1 ? 2 : 1;
```
Comment (eric-haibin-lin) [inline src/operator/nn/softmax-inl.h]: I'd suggest we have an env var to trigger stable reduction instead of relying on dtype argument, including other ops like norm

**Comment by pinaraws** [comment]
@nswamy Please address comments

**Comment by pinaraws** [comment]
@mxnet-label-bot add[pr-awaiting-response]

**Comment by pinaraws** [comment]
@mxnet-label-bot remove[pr-awaiting-review]

File: src/operator/nn/softmax-inl.h (hunk 2 above)
Comment (eric-haibin-lin) [reply src/operator/nn/softmax-inl.h]: https://github.com/apache/incubator-mxnet/pull/14830/files

**Comment by eric-haibin-lin** [comment]
Fixed in https://github.com/apache/incubator-mxnet/pull/15037
